lib/spack/spack/cmd/install.py

```diff
@@ -283,36 +283,41 @@ def install(parser, args, **kwargs):
     reporter.specs = specs
     with reporter:
         if args.overwrite:
-            # If we asked to overwrite an existing spec we must ensure that:
-            # 1. We have only one spec
-            # 2. The spec is already installed
-            assert len(specs) == 1, \
-                "only one spec is allowed when overwriting an installation"
-
-            spec = specs[0]
-            t = spack.store.db.query(spec)
-            assert len(t) == 1, "to overwrite a spec you must install it first"
-
-            # Give the user a last chance to think about overwriting an already
-            # existing installation
-            if not args.yes_to_all:
-                tty.msg('The following package will be reinstalled:\n')
 
+            installed = filter(lambda x: x,
```

Comment:
**TL;DR** I think you need to wrap this into a call to `list` for the code to be correct also in Python 3. Another thing you might want to do (but that's just style as far as I am concerned) is to use `filter(None, iterable)` as `None` implies using the identity function according to the docs.

I need to check by actually trying the PR, but I suspect you have a bug in Python 3 here. One of the breaking changes in Python 3 was that `filter` and `map` became lazy and return a generator object. This means that:
1. Something that generates an empty list, if tested, is True-ish
2. Once you iterate over the generator it's consumed